migrations/kdvh/dump.go

```diff
@@ -0,0 +1,244 @@
+package kdvh
+
+import (
+	"database/sql"
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	_ "github.com/jackc/pgx/v5/stdlib"
+
+	"migrate/utils"
+)
+
+type DumpConfig struct {
+	BaseDir     string   `short:"p" long:"path" default:"./dumps/kdvh" description:"Location the dumped data will be stored in"`
+	TablesCmd   string   `short:"t" long:"table" default:"" description:"Optional comma separated list of table names. By default all available tables are processed"`
+	StationsCmd string   `short:"s" long:"stnr" default:"" description:"Optional comma separated list of stations IDs. By default all station IDs are processed"`
+	ElementsCmd string   `short:"e" long:"elem" default:"" description:"Optional comma separated list of element codes. By default all element codes are processed"`
+	Overwrite   bool     `long:"overwrite" description:"Overwrite any existing dumped files"`
+	Email       []string `long:"email" description:"Optional email address used to notify if the program crashed"`
+
+	Tables   []string
+	Stations []string
+	Elements []string
+}
+
+func (config *DumpConfig) setup() {
+	if config.TablesCmd != "" {
+		config.Tables = strings.Split(config.TablesCmd, ",")
+	}
+	if config.StationsCmd != "" {
+		config.Stations = strings.Split(config.StationsCmd, ",")
+	}
+	if config.ElementsCmd != "" {
+		config.Elements = strings.Split(config.ElementsCmd, ",")
+	}
+}
+
+func (config *DumpConfig) Execute([]string) error {
+	config.setup()
+
+	conn, err := sql.Open("pgx", os.Getenv("KDVH_PROXY_CONN"))
+	if err != nil {
+		slog.Error(err.Error())
+		return nil
+	}
+
+	for _, table := range KDVH {
+		if config.Tables != nil && !slices.Contains(config.Tables, table.TableName) {
+			continue
+		}
+		table.Dump(conn, config)
+	}
+
+	return nil
+}
+
+func (table *Table) Dump(conn *sql.DB, config *DumpConfig) {
+	defer utils.SendEmailOnPanic(fmt.Sprintf("%s dump", table.TableName), config.Email)
+
+	table.Path = filepath.Join(config.BaseDir, table.Path)
+	if _, err := os.ReadDir(table.Path); err == nil && !config.Overwrite {
+		slog.Info(fmt.Sprint("Skipping data dump of ", table.TableName, " because dumped folder already exists"))
+		return
+	}
+
+	utils.SetLogFile(table.TableName, "dump")
+	elements, err := table.getElements(conn, config)
+	if err != nil {
+		return
+	}
+
+	bar := utils.NewBar(len(elements), table.TableName)
+
+	// TODO: should be safe to spawn goroutines/waitgroup here with connection pool?
+	bar.RenderBlank()
+	for _, element := range elements {
+		table.dumpElement(element, conn, config)
+		bar.Add(1)
+	}
+}
+
+// TODO: maybe we don't do this? Or can we use pgdump/copy?
+// The problem is that there are no indices on the tables, that's why the queries are super slow
+// Dumping the whole table might be a lot faster (for T_MDATA it's ~10 times faster!),
```

Comment:
Yep, there are problems with this approach too:
1) bloat, there are a lot of NULL entries
2) connection problems could be annoying, you would need to redump from scratch basically. Not ideal for big tables
3) postprocessing, you need to loop over the data and flag dumps and join them manually (instead of letting the DB do the job for you)

Yesterday Ketil pointed out that we can use the ELEM tables (each row stores metadata for a single timeseries) to get the stations available in the DATA and FLAG tables. I did some tests, and it's indeed much faster, so I'm inclined to go with this approach.

Reply:
Ah, that makes sense. How long are we talking about this taking for the dumps? I think importing is more time-sensitive for us than dumping, so as long as the dump isn't taking days to complete I'm happy to sacrifice some speed there for convenience and reliability

Reply:
It could be up to a day for big tables, but hopefully it's a single time thing :+1: